File: src/deadlock_build_sync/purchase_planner.py
"""Replan optional purchases from actual inventory without selling core items."""

from __future__ import annotations

from collections import Counter
from typing import TYPE_CHECKING

from .mechanics import (
    BASE_INVENTORY_SLOTS,
    InventoryState,
    MechanicsError,
    purchase_item,
)
from .purchase_guidance_types import PurchasePlan, PurchaseState, PurchaseStep

if TYPE_CHECKING:
    from .mechanics import ItemGraph


def covered(graph: ItemGraph, item: int, owned: tuple[int, ...]) -> bool:
    return item in owned or any(
        item in graph.transitive_components(parent) for parent in owned
    )


class PurchasePlanner:
    def __init__(self, graph: ItemGraph, owned: tuple[int, ...], flex: int) -> None:
        if isinstance(flex, bool) or not isinstance(flex, int):
            raise MechanicsError("Flex slots must be an integer")
        self.graph = graph
        self.state = InventoryState(owned, flex)
        for item, count in Counter(owned).items():
            if isinstance(item, bool) or not isinstance(item, int):
                raise MechanicsError("Owned item IDs must be integers")
            node = graph.require(item)
            if count > node.max_count or (node.unique and count > 1):
                raise MechanicsError("Inventory exceeds an item's ownership limit")
        if len(owned) > BASE_INVENTORY_SLOTS + flex:
            raise MechanicsError("Current inventory exceeds available slots")
        if sum(graph.require(item).active for item in owned) > 4:
            raise MechanicsError("Current inventory exceeds four active bindings")
        self.actions: list[PurchaseStep] = []
        self.total = 0

    def acquire(self, item: int, *, exact: bool = False) -> None:
        satisfied = (
            item in self.state.owned
            if exact
            else covered(self.graph, item, self.state.owned)
        )
        if satisfied:
            return
        for component in self.graph.components[item]:
            self.acquire(component, exact=True)
        cash = self.graph.incremental_cash_cost(item, self.state.owned)
        consumed = tuple(
            part for part in self.graph.components[item] if part in self.state.owned
        )
        self.state = purchase_item(self.graph, self.state, item)
        self.total += cash
        self.actions.append(
            PurchaseStep(
                item,
                self.graph.require(item).name,
                cash,
                self.total,
                consumed,
                self.state.owned,
            )
        )

    def result(self, liquid_souls: int | None) -> PurchasePlan:
        shortage = (
            max(0, self.actions[0].incremental_cost - liquid_souls)
            if self.actions and liquid_souls is not None
            else None
        )
        decision = (
            "complete"
            if not self.actions
            else "save"
            if shortage
            else "buy"
            if liquid_souls is not None
            else "check cash"
        )
        return PurchasePlan(
            tuple(self.actions), self.state.owned, self.total, decision, shortage
        )


def validate_positions(
    graph: ItemGraph,
    path: tuple[int, ...],
    positions: dict[int, int],
    owned: tuple[int, ...],
) -> None:
    for item, index in positions.items():
        if isinstance(item, bool) or not isinstance(item, int):
            raise TypeError("Selected item IDs must be integers")
        if (
            isinstance(index, bool)
            or not isinstance(index, int)
            or not 0 <= index <= len(path)
        ):
            raise ValueError(
                "Purchase position must be an integer within the default path"
            )
        components = graph.transitive_components(item)
        remaining = {
            value for value in path[index:] if not covered(graph, value, owned)
        }
        if remaining.intersection(components):
            raise ValueError("Purchase position precedes a required core item")
        if any(
            component in positions and positions[component] > index
            for component in components
        ):
            raise ValueError("Selected upgrade precedes its selected component")


def plan_purchases(
    graph: ItemGraph,
    path: tuple[int, ...],
    core: tuple[int, ...],
    positions: dict[int, int],
    *,
    state: PurchaseState | None = None,
) -> PurchasePlan:
    """Return a legal core-preserving route, including consumed-component rebuys.

    Returns:
        The full remaining route and the affordable next action.

    Raises:
        ValueError: If a position or liquid soul value is invalid.
        MechanicsError: If inventory or the resulting path is illegal.

    """
    state = state or PurchaseState()
    liquid_souls = state.liquid_souls
    if liquid_souls is not None and (
        isinstance(liquid_souls, bool)
        or not isinstance(liquid_souls, int)
        or liquid_souls < 0
    ):
        raise ValueError("Liquid souls must be a nonnegative integer")
    planner = PurchasePlanner(graph, state.owned, state.flex)
    validate_positions(graph, path, positions, state.owned)
    selected = sorted(
        positions,
        key=lambda item: (
            positions[item],
            len(graph.transitive_components(item)),
            item,
        ),
    )
    for index in range(len(path) + 1):
        for item in selected:
            if positions[item] == index:
                planner.acquire(item)
        if index < len(path):
            planner.acquire(path[index])
    if not all(covered(graph, item, planner.state.owned) for item in core):
        raise MechanicsError("Choice abandons the core's upgrade lineages")
    return planner.result(liquid_souls)

File: crates/deadlock-guides/src/purchase_planner.rs
use std::collections::BTreeMap;

use deadlock_data::{Error, Result};
use deadlock_input::ItemGraph;

use crate::inventory::{InventoryState, MAX_ACTIVE_ITEMS, purchase_item};
use crate::purchase_plan_types::{PurchasePlan, PurchaseState, PurchaseStep};
use crate::purchase_route::{
    is_item_or_upgrade_owned, resolve_route_targets, validate_purchase_positions,
};

#[derive(Debug)]
struct PurchasePlanner<'graph> {
    graph: &'graph ItemGraph,
    state: InventoryState,
    actions: Vec<PurchaseStep>,
    total_cost: u64,
}

impl<'graph> PurchasePlanner<'graph> {
    fn new(graph: &'graph ItemGraph, state: &PurchaseState) -> Result<Self> {
        let state = InventoryState::new(state.owned.clone(), state.flex)?;
        validate_owned_items(graph, &state)?;
        Ok(Self {
            graph,
            state,
            actions: Vec::new(),
            total_cost: 0,
        })
    }

    fn add_item_purchases(&mut self, item_id: u64, exact: bool, depth: usize) -> Result<()> {
        if depth > 64 || self.actions.len() >= 4096 {
            return Err(Error::new(
                "Purchase plan exceeds its depth or action limit",
            ));
        }
        let satisfied = if exact {
            self.state.owned().contains(&item_id)
        } else {
            is_item_or_upgrade_owned(self.graph, item_id, self.state.owned())?
        };
        if satisfied {
            return Ok(());
        }
        for component in self.graph.components(item_id)?.to_vec() {
            self.add_item_purchases(component, true, depth + 1)?;
        }
        let incremental_cost = self
            .graph
            .incremental_cash_cost(item_id, self.state.owned())?;
        let consumed_items = self
            .graph
            .components(item_id)?
            .iter()
            .copied()
            .filter(|id| self.state.owned().contains(id))
            .collect();
        self.state = purchase_item(self.graph, &self.state, item_id, 0)?;
        self.total_cost = self
            .total_cost
            .checked_add(incremental_cost)
            .ok_or_else(|| Error::new("Purchase plan cost exceeds 64 bits"))?;
        self.actions.push(PurchaseStep {
            item_id,
            name: self.graph.require(item_id)?.name.clone(),
            incremental_cost,
            cumulative_cost: self.total_cost,
            consumed_items,
            owned_after: self.state.owned().to_vec(),
        });
        Ok(())
    }

    fn build_plan(self, liquid_souls: Option<u64>) -> PurchasePlan {
        let shortage = self
            .actions
            .first()
            .zip(liquid_souls)
            .map(|(action, souls)| action.incremental_cost.saturating_sub(souls));
        let decision = if self.actions.is_empty() {
            "complete"
        } else if shortage.is_some_and(|amount| amount > 0) {
            "save"
        } else if liquid_souls.is_some() {
            "buy"
        } else {
            "check cash"
        };
        PurchasePlan {
            actions: self.actions,
            final_inventory: self.state.owned().to_vec(),
            remaining_cost: self.total_cost,
            decision: decision.into(),
            save_souls: shortage,
        }
    }
}

fn validate_owned_items(graph: &ItemGraph, state: &InventoryState) -> Result<()> {
    let mut counts = BTreeMap::<u64, u32>::new();
    let mut active = 0;
    for id in state.owned() {
        let item = graph.require(*id)?;
        let count = counts.entry(*id).or_default();
        *count += 1;
        if *count > item.max_count || (item.unique && *count > 1) {
            return Err(Error::new("Inventory exceeds an item ownership limit"));
        }
        active += usize::from(item.active);
    }
    if active > MAX_ACTIVE_ITEMS {
        return Err(Error::new("Current inventory exceeds four active bindings"));
    }
    Ok(())
}

/// # Errors
/// Returns an error when the current inventory or exact item purchase violates the component or inventory rules.
pub fn plan_exact_item_purchase(
    graph: &ItemGraph,
    item: u64,
    state: &PurchaseState,
) -> Result<PurchasePlan> {
    let mut planner = PurchasePlanner::new(graph, state)?;
    planner.add_item_purchases(item, true, 0)?;
    Ok(planner.build_plan(state.liquid_souls))
}

/// Calculates remaining purchases from the current inventory and available souls.
///
/// # Errors
/// Returns an error when a purchase position, inventory, component path, or resulting core is invalid.
pub fn plan_purchases(
    graph: &ItemGraph,
    path: &[u64],
    core: &[u64],
    positions: &BTreeMap<u64, usize>,
    state: &PurchaseState,
) -> Result<PurchasePlan> {
    let mut planner = PurchasePlanner::new(graph, state)?;
    validate_purchase_positions(graph, path, positions, &state.owned)?;
    let mut selected = positions
        .iter()
        .map(|(item, position)| Ok((*position, graph.transitive_components(*item)?.len(), *item)))
        .collect::<Result<Vec<_>>>()?;
    selected.sort_unstable();
    let targets = resolve_route_targets(graph, path)?;
    let checkpoints = path.iter().zip(&targets).map(Some).chain([None]);
    for (index, checkpoint) in checkpoints.enumerate() {
        for (position, _, item) in &selected {
            if *position == index {
                planner.add_item_purchases(*item, false, 0)?;
            }
        }
        if let Some((item, target)) = checkpoint
            && !is_item_or_upgrade_owned(graph, *target, planner.state.owned())?
        {
            planner.add_item_purchases(*item, item != target, 0)?;
        }
    }
    for item in core {
        if !is_item_or_upgrade_owned(graph, *item, planner.state.owned())? {
            return Err(Error::new(
                "Purchase plan does not retain all required core items",
            ));
        }
    }
    Ok(planner.build_plan(state.liquid_souls))
}
